File: Ruby/problem05.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

class Proc
  def self_curry
    self.curry.call(self)
  end
end


def problem05(nrlist, oplist, answer)
  if nrlist.length == 0
    return []
  else
    exprlist = []
    lambda {|f, expr, i|
      if i < nrlist.length
        oplist.each {|op|
          f.call(f, expr + op + nrlist[i].to_s, i + 1)
        }
      elsif eval(expr) == answer
        exprlist.push(expr)
      end
    }.self_curry.call(nrlist[0].to_s, 1)
    return exprlist
  end
end


if __FILE__ == $0
  ANSWER = 100
  problem05((1..9).to_a, [' + ', ' - ', ''], ANSWER).each {|expr|
    puts expr + ' = ' + ANSWER.to_s
  }
end

File: JavaScript/problem05.js
;(function(print) {
  'use strict';

  function problem05(nrlist, oplist, answer) {
    if (nrlist.length === 0) {
      return [];
    } else {
      var exprlist = [];
      (function _self(expr, i) {
        if (i < nrlist.length) {
          oplist.forEach(function(op) {
            _self(expr + op + nrlist[i], i + 1);
          });
        } else if (eval(expr) == answer) {
          exprlist.push(expr);
        }
      })('' + nrlist[0], 1);
      return exprlist;
    }
  }

  var ANSWER = 100;
  problem05([1, 2, 3, 4, 5, 6, 7, 8, 9], [' + ', ' - ', ''], ANSWER).forEach(function(e) {
    print(e + ' = ' + ANSWER);
  });
})(typeof console === 'undefined' ? print : console.log);
